### ISSUE TITLE
Algebraic solve with multigrid crashes after a variational solve

### ISSUE THREAD

**Issue by salazardetroya** [firedrakeproject/firedrake#1578]
I found this bug using pyadjoint with multigrid options, where the adjoint solve uses firedrake algebraic solve.

```python
from firedrake import *

mesh = UnitSquareMesh(8, 8)
hierarchy = MeshHierarchy(mesh, 1)
mesh = hierarchy[-1]

V = FunctionSpace(mesh, "CG", 1)

u = TrialFunction(V)
v = TestFunction(V)

a = dot(grad(u), grad(v))*dx

bcs = DirichletBC(V, zero(), (1, 2, 3, 4))

x, y = SpatialCoordinate(mesh)

f = Constant(0.1)

L = f*v*dx

u = Function(V)

parameters = {
   "ksp_type": "cg",
   "ksp_converged_reason": None,
   "pc_type": "mg",
}

A = assemble(a, bcs=bcs)
b = assemble(L, bcs=bcs)
solve(a == L, u, bcs=bcs, solver_parameters=parameters)
solve(a == L, u, bcs=bcs, solver_parameters=parameters)
solve(A, u.vector(), b, solver_parameters=parameters)
```

I call the variational solve twice to check for crash there, but no problem. Only the third solve, the algebraic, crashes. If I leave only the algebraic solve, it works well. 

Related question, how does the Firedrake team go about debugging bugs like this? I tried using gdb, but I need python-dbg for the debugging symbols. Setting up firedrake venv with python-dbg is a bit tricky though.

**Comment by wence-** [comment]
Hmm, I wonder what data we've failed to leave lying around.

Normally I debug this stuff with a combination of valgrind and gdb. Normally the crashes aren't really in Python. If they're in generated code you can first `export PYOP2_DEBUG=1` to get debug symbols in that compiled code.

I suspect we're leaving something hanging around on DMs somewhere.

**Comment by salazardetroya** [comment]
Out of curiosity, have you tried setting up Firedrake with python-dbg? What about using cygdb for the cython code?

**Comment by wence-** [comment]
I've never tried it. My crystal ball brief debugging suggested that the problem was in a cleanup failure inside PETSc (basically accessing a pointer that had already been freed).

**Comment by salazardetroya** [comment]
Hello, is there any update on this issue? Fixing it would allow pyadjoint users use geometric multigrid and solve large problems.

**Comment by salazardetroya** [comment]
This bug was actually fixed with #1735 ! thanks !
